### PR TITLE
Recursion

### PR DESCRIPTION
**Pull request by finico**
Recursive functions
And I reduced number of test selection_sort repeats to 10 to increase speed

### CODE REVIEW

File: src/recursive_len/__init__.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+def recursive_len(arr):
+    if arr:
+        return 1 + recursive_len(arr[1:])
+    return 0
```
Comment (codewylle) [inline src/recursive_len/__init__.py]: else?

File: src/recursive_max/__init__.py [hunk 2]
```diff
@@ -0,0 +1,7 @@
+def recursive_max(arr):
+    if not arr:
+        return None
```
Comment (codewylle) [inline src/recursive_max/__init__.py]: Throwing error is better. What is maximum of nothing?

File: src/recursive_max/__init__.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+    if len(arr) == 1:
+        return arr[0]
+    x = recursive_max(arr[1:])
+    return arr[0] if arr[0] > x else x
```
Comment (codewylle) [inline src/recursive_max/__init__.py]: `head = arr[0]`
and maybe `tail = arr[1:]`

File: src/recursive_sum/__init__.py [hunk 4]
```diff
@@ -0,0 +1,6 @@
+def recursive_sum(arr):
+    if not arr:
+        return 0
+    if len(arr) == 1:
```
Comment (codewylle) [inline src/recursive_sum/__init__.py]: it is unnecessary, previous case is sufficient

File: tests/test_recursive_countdown.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
...
+from unittest.mock import patch
+
+from src.recursive_countdown import recursive_countdown
+
+
+class TestRecursiveCountdown(TestCase):
+    @patch('builtins.print')
+    def test_countdown(self, mocked_print):
+        recursive_countdown(3)
+        self.assertEqual(len(mocked_print.mock_calls), 4)
```
Comment (codewylle) [inline tests/test_recursive_countdown.py:11]: check arguments?

File: tests/test_recursive_len.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+import unittest
+
+from src.recursive_len import recursive_len
+
+
+class TestRecursiveLen(unittest.TestCase):
+    def test_len(self):
+        self.assertEqual(recursive_len([1, 2, 3, 4, 5]), 5)
```
Comment (codewylle) [inline tests/test_recursive_len.py]: check `recursive_len([])`

File: tests/test_recursive_max.py [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+
+class TestRecursiveMax(unittest.TestCase):
+    def test_when_empty(self):
+        self.assertIsNone(recursive_max([]))
+
+    def test_max_pair(self):
+        self.assertEqual(recursive_max([5, 2]), 5)
+
+    def test_max(self):
+        self.assertEqual(recursive_max([1, 55, 29, 73, 6]), 73)
```
Comment (codewylle) [inline tests/test_recursive_max.py]: check array containing one value

File: tests/test_recursive_max.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
+import unittest
+
+from src.recursive_max import recursive_max
+
+
+class TestRecursiveMax(unittest.TestCase):
+    def test_when_empty(self):
+        self.assertIsNone(recursive_max([]))
+
+    def test_max_pair(self):
```
Comment (codewylle) [inline tests/test_recursive_max.py]: is it needed when there is `test_max`? for debugging?

File: src/recursive_len/__init__.py (hunk 1 above)
Comment (finico) [reply src/recursive_len/__init__.py]: No. But I can do it through ternary operator

File: src/recursive_max/__init__.py (hunk 2 above)
Comment (finico) [reply src/recursive_max/__init__.py]: ok, I'll get it

File: src/recursive_sum/__init__.py (hunk 4 above)
Comment (finico) [reply src/recursive_sum/__init__.py]: No Man. Previous case checks for a list emptiness

File: tests/test_recursive_max.py (hunk 8 above)
Comment (finico) [reply tests/test_recursive_max.py]: No, I replaced it with the check for a list with one value

File: tests/test_recursive_len.py (hunk 6 above)
Comment (finico) [reply tests/test_recursive_len.py]: done

File: tests/test_recursive_countdown.py (hunk 5 above)
Comment (finico) [reply tests/test_recursive_countdown.py:11]: done

File: tests/test_recursive_max.py (hunk 7 above)
Comment (finico) [reply tests/test_recursive_max.py]: done

File: src/recursive_len/__init__.py (hunk 1 above)
Comment (codewylle) [reply src/recursive_len/__init__.py]: You are right about ternary if

File: src/recursive_sum/__init__.py (hunk 4 above)
Comment (codewylle) [reply src/recursive_sum/__init__.py]: If this case will be absent and arr's len will be 1, next case happens. Let reduce it: `return arr[0] + recursive_sum(arr[1:])` => `return arr[0] + recursive_sum([])` => `return arr[0] + 0` => `return arr[0]`. Bingo! They are identical
You can investigate various FP libraries and you will see coders excludes superfluous cases can be infered from other cases

Comment (finico) [reply src/recursive_sum/__init__.py]: Ah! I got it

Comment (codewylle) [reply src/recursive_sum/__init__.py]: Sorry, man, my first comment was inaccurate, it confused you

File: src/recursive_max/__init__.py [hunk 9]
```diff
@@ -0,0 +1,8 @@
+def recursive_max(arr):
+    assert arr, "List can not be empty"
+
+    head = arr[0]
+    if len(arr) == 1:
+        return head
+    x = recursive_max(arr[1:])
+    return head if head > x else x
```
Comment (codewylle) [inline src/recursive_max/__init__.py]: `head` & `x` is incoherent names, peoples usually use `head` & `tail` or `x` & `xs`

**Review by codewylle — APPROVED** [review]

File: src/recursive_max/__init__.py (hunk 9 above)
Comment (codewylle) [reply src/recursive_max/__init__.py]: I was inattentive, it isn't tail :( Sorry, man, name it `maxOther` or alike

Comment (finico) [reply src/recursive_max/__init__.py]: I've renamed it to `guess_max`

**Review by codewylle — APPROVED** [review]

**Pull request merged**